### PR TITLE
Timetables data inserter

### PR DESCRIPTION
**Pull request by Leitsi**
Resolves https://github.com/HSLdevcom/jore4/issues/1333

- - -
This change is [Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192)

### CODE REVIEW

**Review by HenrikHartiala — CHANGES_REQUESTED** [review]
> Reviewed 46 of 66 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192)* status: 46 of 66 files reviewed, 6 unresolved discussions (waiting on @Leitsi)

___
*[`test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts` line 95 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOk2up8diyBn8TjPyR:-NbOk2up8diyBn8TjPyS:bfzs583) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts#L95)):*
> ```XML
>             vehicle_schedule_frame_id: null,
>             substitute_operating_day_by_line_type_id:
>               datasetHelper.getSubstituteOperatingDayByLineTypes(
> ```

A thought raised from this for possible future iterations: should we type these? It would add a bit more boilerplate, but would also give autocompletion.
So we could have some exported enum&/record etc. which has all the different different choices.

Definitely not required for this iteration, if decided to have this, but to open discussion. 
What do you think?



_Suggestion:_
```XML
              datasetHelper.getSubstituteOperatingDayByLineTypes(
                substituteOperatingDayByLineTypeByNames.saturday20230520,
              ).substitute_operating_day_by_line_type_id,
```
___
*[`test/hasura/timetables-data-inserter/generic/table-data.ts` line 24 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOaXjLAGwR7B9J3qq9:-NbOaXjLAGwR7B9J3qqA:b-oqt04k) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/table-data.ts#L24)):*
> ```XML
>   );
>   const vehicleServices = vehicleScheduleFrames
>     .map((vsf) => Object.values(vsf._vehicle_services || {}))
> ```

These could also be flatMaps, but either is fine for me.



_Suggestion:_
```XML
    .flatMap((vsf) => Object.values(vsf._vehicle_services || {}))
```
___
*[`test/hasura/timetables-data-inserter/generic/timetables-dataset-helper.ts` line 9 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOctEH9CGhKulPeXC3:-NbOctEH9CGhKulPeXC4:b7b8ozm) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/timetables-dataset-helper.ts#L9)):*
> ```XML
> } from './types';
> 
> export const sanityCheckLabelGroupResult = (
> ```

This could benefit from a minor JSDoc description. 
___
*[`test/hasura/timetables-data-inserter/generic/models/vehicle-journey.ts` line 20 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOdi_o2fwT8437WXbl:-NbOdi_o2fwT8437WXbm:b-dm6j3r) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/models/vehicle-journey.ts#L20)):*
> ```XML
> 
>   const journeyPatternRefName =
>     vjWithId._journey_pattern_ref_name || 'TODO: how to handle this';
> ```

Should this TODO be resolved? Or could this have something more descriptive about the TODO.
___
*[`test/hasura/timetables-data-inserter/hsl/example.json` line 1 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOfRlyBNQWOlHo-Ah2:-NbOfRlz5wOtuO-sCXW8:b-6e84mc) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/example.json#L1)):*
> ```JSON
> {
> ```

This example json is very nice :+1: 
___
*[`test/hasura/timetables-data-inserter/hsl/table-data.ts` line 28 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOgamv6tHqbGWl27vB:-NbOgamv6tHqbGWl27vC:be6u66z) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/table-data.ts#L28)):*
> ```XML
>   );
>   const substituteOperatingDayByLineTypeIds = substituteOperatingPeriods
>     .map((period) =>
> ```

Could also be flatMap
___
*[`test/hasura/timetables-data-inserter/hsl/timetables-dataset-helper.ts` line 19 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOh53eE_VJsoUuvkAL:-NbOh53eE_VJsoUuvkAM:b-2gne6d) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/timetables-dataset-helper.ts#L19)):*
> ```XML
>   const substituteOperatingDayByLineTypesByLabel =
>     flattened.substituteOperatingPeriods
>       .map((sop) => Object.entries(sop._substitute_operating_day_by_line_types))
> ```

same same
___
*[`test/hasura/timetables-data-inserter/hsl/models/substitute-operating-period.ts` line 10 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOhZPC98dr5eRzOMXe:-NbOhZPC98dr5eRzOMXf:bhqvzaa) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/models/substitute-operating-period.ts#L10)):*
> ```XML
>   substituteOperatingPeriod: SubstituteOperatingPeriodInput,
> ): SubstituteOperatingPeriodOutput => {
>   // const result = assignId(substituteOperatingPeriod, ...);
> ```

Big NIT: maybe add `TODO: ` in front. I was about to comment that we have commented code here :D 
___
*[`test/hasura/timetables-data-inserter/hsl/models/vehicle-schedule-frame.ts` line 27 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOiLhg-QNc6Tph55PJ:-NbOiLhg-QNc6Tph55PK:b-8uuamb) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/models/vehicle-schedule-frame.ts#L27)):*
> ```XML
>   return {
>     booking_label: buildName(vehicleScheduleFrame as EntityName).fi_FI,
>     booking_description_i18n: undefined,
> ```

Is this here for clarence/readability, or why do we need to explicitly say it's undefined?
___
*[`test/hasura/util/setup.ts` line 16 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOj4VE-thSz9pwdKGw:-NbOj4VE-thSz9pwdKGx:b-clfvys) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/util/setup.ts#L16)):*
> ```XML
>   return promiseSequence(
>     tableData
>       .filter((table) => table.data.length) // Crashes with empty insert queries.
> ```

NIT, If I understood this right, this crashes without the filtering? If so, then I'd rephrase the comment a bit, because now I was first thinking that is this a TODO / should this crash be handled somehow :D.

If this was not what it meant, then should we handle the crash?



_Suggestion:_
```XML
      .filter((table) => table.data.length) // We need to filter to avoid crash with empty insert queries.
```

**Review by Leitsi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192)* status: 46 of 66 files reviewed, 3 unresolved discussions (waiting on @HenrikHartiala)

___
*[`test/hasura/timetables-data-inserter/generic/table-data.ts` line 24 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOaXjLAGwR7B9J3qq9:-NbPBIb3AXoiCZ5y3Nig:b-3dezrd) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/table-data.ts#L24)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

These could also be flatMaps, but either is fine for me.
</blockquote></details>

Right, this exists. Changed. :thumbsup: 
___
*[`test/hasura/timetables-data-inserter/generic/timetables-dataset-helper.ts` line 9 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOctEH9CGhKulPeXC3:-NbP9NFhAIFgWPFC_J36:b9y268c) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/timetables-dataset-helper.ts#L9)):*
Added a short comment.  Not too critical though, since this is not intended to be used eg. from tests: exported only for the hsl side (also explained this in the comment).

And, slapped a `@protected` annotation here. Seems accurate enough?
https://jsdoc.app/tags-protected.html
> Typically, this tag indicates that a symbol is only available, or should only be used, **within the current module**.
___
*[`test/hasura/timetables-data-inserter/generic/models/vehicle-journey.ts` line 20 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOdi_o2fwT8437WXbl:-NbP8F177f0jBVjUWGcr:blai60s) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/models/vehicle-journey.ts#L20)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

Should this TODO be resolved? Or could this have something more descriptive about the TODO.
</blockquote></details>

Resolved by removing it. The `_journey_pattern_ref_name` is required anyways.
Also refactored this bit slightly, added explicit error throw for case when `journeyPatternRef` is not found.
___
*[`test/hasura/timetables-data-inserter/hsl/example.json` line 1 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOfRlyBNQWOlHo-Ah2:-NbP8-RHBFqMzLdRIX4H:bi9estx) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/example.json#L1)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

This example json is very nice :+1: 
</blockquote></details>

:pray: 
___
*[`test/hasura/timetables-data-inserter/hsl/models/substitute-operating-period.ts` line 10 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOhZPC98dr5eRzOMXe:-NbP7ugzBp0Wm5qxNHGu:b4c9lyr) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/models/substitute-operating-period.ts#L10)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

Big NIT: maybe add `TODO: ` in front. I was about to comment that we have commented code here :D 
</blockquote></details>

:thumbsup: 
___
*[`test/hasura/timetables-data-inserter/hsl/models/vehicle-schedule-frame.ts` line 27 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOiLhg-QNc6Tph55PJ:-NbP7htlCX_ZFs_rHfv5:b-go1w3r) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/models/vehicle-schedule-frame.ts#L27)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

Is this here for clarence/readability, or why do we need to explicitly say it's undefined?
</blockquote></details>

Doesn't seem to be necessary, removing.
Think this was needed at some point when I didn't have output type specified explicitly.
___
*[`test/hasura/util/setup.ts` line 16 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOj4VE-thSz9pwdKGw:-NbP6zYb5VcrGfBLiFg1:b-6ugp6f) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/util/setup.ts#L16)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

NIT, If I understood this right, this crashes without the filtering? If so, then I'd rephrase the comment a bit, because now I was first thinking that is this a TODO / should this crash be handled somehow :D.

If this was not what it meant, then should we handle the crash?
</blockquote></details>

Rephrased. Couldn't think of a neater fix for this (not like the insert would do anything anyway with empty input, so might as well skip it altogether).

**Review by Leitsi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192)* status: 39 of 67 files reviewed, 4 unresolved discussions (waiting on @HenrikHartiala and @Leitsi)

___
*[`test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts` line 95 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOk2up8diyBn8TjPyR:-NbPDkA-9ZQ6iHmptaCd:bvrhleg) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts#L95)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

A thought raised from this for possible future iterations: should we type these? It would add a bit more boilerplate, but would also give autocompletion.
So we could have some exported enum&/record etc. which has all the different different choices.

Definitely not required for this iteration, if decided to have this, but to open discussion. 
What do you think?
</blockquote></details>

I spent a great deal of time trying to get the built dataset to retain typed keys, so autocomplete would've worked. Couldn't get that working at least.

This idea feels like we might run into similar problems.
At least those records should work automatically or be created via some helper function, and not be manually maintained. I'll see if I can cook up something.

**Review by Leitsi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192)* status: 39 of 67 files reviewed, 3 unresolved discussions (waiting on @HenrikHartiala)

___
*[`test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts` line 95 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOk2up8diyBn8TjPyR:-NbTrOX-Eqw-Ziswl9wj:b-ty35kx) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts#L95)):*
<details><summary><i>Previously, Leitsi (Tommi Leinamo) wrote…</i></summary><blockquote>

I spent a great deal of time trying to get the built dataset to retain typed keys, so autocomplete would've worked. Couldn't get that working at least.

This idea feels like we might run into similar problems.
At least those records should work automatically or be created via some helper function, and not be manually maintained. I'll see if I can cook up something.
</blockquote></details>

Spent some time investigating but couldn't get any automatic solution working. Won't add for now.

**Review by HenrikHartiala — APPROVED** [review]
Absolutely great work! This will help us with the data insertion a lot. I checkouted the code, checked some bits from here and there and did a minor testing myself. Everything seemed way clearer than the previous system.
Since this is a big PR, and a lot of stuff is happening, I did not dive in to every detail. Tests work and the inserter works nicely, so if there are bugs or enhancements, they are so much easier to do by iteration in their individual PRS. So I think this is good to go!

> Reviewed 1 of 66 files at r1, 2 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-:-NbUNpMrAKmk7kmp7XXY:b-5glxod)* status: 42 of 67 files reviewed, 1 unresolved discussion (waiting on @Leitsi)

___
*[`test/hasura/hsl/timetablesdb/datasets/additional-sets/timetables/specialAprilFools2023Dataset.ts` line 47 at r2](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbUSfxu6xW4waQ3UUmv:-NbUSfxu6xW4waQ3UUmw:b-jah9oy) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/ecf23fba811d4002f56f02ed64d7ae7812115141/test/hasura/hsl/timetablesdb/datasets/additional-sets/timetables/specialAprilFools2023Dataset.ts#L47)):*
> ```XML
> };
> 
> export const specialAprilFools2023Dataset: TableData<HslTimetablesDbTables>[] =
> ```

Can't remember if you already said some reason to retain these, but if not then by my observation these can now be removed, since they are not used
___
*[`test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts` line 95 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOk2up8diyBn8TjPyR:-NbUBXygAt4f2jTYSDXz:b7d11s8) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/hsl/timetablesdb/integration-tests/function-get_timetables_and_substitute_operating_days/get-timetables-and-substitute-operating-days.spec.ts#L95)):*
<details><summary><i>Previously, Leitsi (Tommi Leinamo) wrote…</i></summary><blockquote>

Spent some time investigating but couldn't get any automatic solution working. Won't add for now.
</blockquote></details>

Yep, discussed about this in slack, and we shall let this idea brew for now.
___
*[`test/hasura/timetables-data-inserter/generic/table-data.ts` line 24 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOaXjLAGwR7B9J3qq9:-NbUBluI-SXGEItO4YWj:b-8fx2u7) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/table-data.ts#L24)):*
<details><summary><i>Previously, Leitsi (Tommi Leinamo) wrote…</i></summary><blockquote>

Right, this exists. Changed. :thumbsup: 
</blockquote></details>

:+1: 
___
*[`test/hasura/timetables-data-inserter/generic/timetables-dataset-helper.ts` line 9 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOctEH9CGhKulPeXC3:-NbUCT6t16SvjOzo_jR8:bppc5eh) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/timetables-dataset-helper.ts#L9)):*
<details><summary><i>Previously, Leitsi (Tommi Leinamo) wrote…</i></summary><blockquote>

Added a short comment.  Not too critical though, since this is not intended to be used eg. from tests: exported only for the hsl side (also explained this in the comment).

And, slapped a `@protected` annotation here. Seems accurate enough?
https://jsdoc.app/tags-protected.html
> Typically, this tag indicates that a symbol is only available, or should only be used, **within the current module**.
</blockquote></details>

Thanks, it's still nice to have a minor description in some cases, even though it is not used elsewhere.

Yeah why not for the protected tag :ok_hand: 
___
*[`test/hasura/timetables-data-inserter/generic/models/vehicle-journey.ts` line 20 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOdi_o2fwT8437WXbl:-NbUCmpn4oxQ7YAcv2eX:b84qvrb) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/generic/models/vehicle-journey.ts#L20)):*
<details><summary><i>Previously, Leitsi (Tommi Leinamo) wrote…</i></summary><blockquote>

Resolved by removing it. The `_journey_pattern_ref_name` is required anyways.
Also refactored this bit slightly, added explicit error throw for case when `journeyPatternRef` is not found.
</blockquote></details>

Thats the best way to resolve things, just remove. :D 

Also +1 to error throw case :
___
*[`test/hasura/timetables-data-inserter/hsl/models/vehicle-schedule-frame.ts` line 27 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOiLhg-QNc6Tph55PJ:-NbUDBmp7aPnAt9AYy5E:b-w1pi47) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/timetables-data-inserter/hsl/models/vehicle-schedule-frame.ts#L27)):*
<details><summary><i>Previously, Leitsi (Tommi Leinamo) wrote…</i></summary><blockquote>

Doesn't seem to be necessary, removing.
Think this was needed at some point when I didn't have output type specified explicitly.
</blockquote></details>

check :+1: 
___
*[`test/hasura/util/setup.ts` line 16 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-NbOj4VE-thSz9pwdKGw:-NbUDFPK8SGdDKdagsFJ:b-giv0vt) ([raw file](https://github.com/HSLdevcom/jore4-hasura/blob/f4670743016a4ed1cf0fe90230165b5a36aa76fd/test/hasura/util/setup.ts#L16)):*
<details><summary><i>Previously, Leitsi (Tommi Leinamo) wrote…</i></summary><blockquote>

Rephrased. Couldn't think of a neater fix for this (not like the insert would do anything anyway with empty input, so might as well skip it altogether).
</blockquote></details>

Thanks, seems more clear now :+1:

**Review by HenrikHartiala — APPROVED** [review]
Ahh, the documentation is also top notch! :tada:  
:lgtm:

> Reviewed 16 of 66 files at r1, 5 of 9 files at r2, 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hasura/192#-:-Nbmw-sp51LaY6C8Et9u:bfrgpui)* status: all files reviewed, 1 unresolved discussion (waiting on @Leitsi)
